Extract the mountpoint code into mountpoint-s3-fs library crate (#1304)
awslabs/mountpoint-s3: mountpoint-s3-fs/README.md

# Mountpoint S3 FS (Unstable)

> ## ⚠️ Warning: Not Production Ready ⚠️
> This library is an implementation of the core functionality of Mountpoint S3 FS. The API interface is **unstable** and subject to **significant changes**. Do not use this library in production environments.

Reviewer: ```suggestion
## ⚠️ Warning: Unstable API ⚠️
This library implements the core functionality of Mountpoint for Amazon S3. The API interface is **unstable** and subject to **significant changes**.
```

Author: Reworded readme

## Overview
Mountpoint for Amazon S3 is a high-throughput file client for mounting an Amazon S3 bucket as a local file system.
This library provides an implementation of key Mountpoint functionality but should be considered as a **work-in-progress**.
We recommend all customers to use the official Mountpoint releases instead: https://github.com/awslabs/mountpoint-s3/releases

## Limitations
- **API Instability:** The API will change soon and will introduce breaking changes.

> - **No Production Support:** At the moment this library is for internal use only.

Reviewer: Remove this line

# Mountpoint S3 FS (Unstable)

## ⚠️ Warning: Not Production Ready ⚠️
This library is an implementation of the core functionality of Mountpoint S3 FS. The API interface is **unstable** and subject to **significant changes**. Do not use this library in production environments.

> ## Overview
> Mountpoint S3 is a high-performance S3-compatible file system designed for efficient cloud storage interactions. This library provides an implementation of key Mountpoint S3 functionality but should be considered as **work in progress**.

Reviewer 2: The action we should give to customers is to go use the full Mountpoint release. I'd also update to match the repository description.

```suggestion
## Overview
Mountpoint for Amazon S3 is a high-throughput file client for mounting an Amazon S3 bucket as a local file system. 

This library provides an implementation of key Mountpoint functionality but should be considered as a **work-in-progress**.
We recommend all customers to use the official Mountpoint releases instead: https://github.com/awslabs/mountpoint-s3/releases
```

## Limitations
- **API Instability:** The API will change soon and will introduce breaking changes.
- **No Production Support:** At the moment this library is for internal use only.

## Usage (Subject to Change)
```rust
#[derive(Parser, Debug)]
#[clap(
    name = "mount-s3",
    about = "Mountpoint for Amazon S3",
    version = build_info::FULL_VERSION,
    group(
        ArgGroup::new("cache_group")
            .multiple(true),
    ),
)]
pub struct AppCliArgs {
    #[clap(flatten)]
    pub cli_args: mountpoint_s3_fs::cli::CliArgs,
}

fn main() -> anyhow::Result<()> {
    let args = AppCliArgs::parse();
    let context = mountpoint_s3_fs::cli::ContextParams {
        full_version: build_info::FULL_VERSION.to_string(),
    };
    mountpoint_s3_fs::cli::main(mountpoint_s3_fs::cli::create_s3_client, args.cli_args, context)
}
}

> ```

Reviewer 2: There's one too many curly braces here. I'd actually recommend not including this in the README anyway. There's an example just from `mountpoint-s3` crate itself, and we can always add a simple, compilable example in `mountpoint-s3-fs/examples/` later if we wish.

Author: Done

## Roadmap
- Extraction of configuration layer

> - Stabilization of the API

Reviewer 2: Drop the roadmap, this isn't something we want to commit to right now.

Author: 1. This is more like a long-term vision
2. I assume we do

Reviewer: Please remove this section

## Contributing

> Please note that the library is still in heavy development. Expect breaking changes and evolving functionality.

Reviewer 2: This isn't really about contributing. Drop this or link here: https://github.com/awslabs/mountpoint-s3/blob/main/doc/CONTRIBUTING.md

## License

> MIT License (or appropriate license).

Reviewer 2: drop?

Author: don't mind
